### ISSUE TITLE
deepspeed微调效果不如单卡ddp

### ISSUE THREAD

**Issue by lionelmessigoat** [InternLM/xtuner#176]
**deepspeed loss只能从2到1，ddp能从2到0.4**

### deepspeed
运行：
`
NPROC_PER_NODE=2 xtuner \
    train config.py \
    --work-dir ./logs \
    --cfg-options train_cfg.max_epochs=1 param_scheduler.T_max=1 \
    --deepspeed xtuner/xtuner/configs/deepspeed/deepspeed_zero2.json `
其中 config.py 配置：https://github.com/lionelmessigoat/xtuner_issue1/blob/main/xtuner_config.py
FireflySFTDataset文件代码：https://github.com/lionelmessigoat/xtuner_issue1/blob/main/firefly_sft.py

###  ddp 使用项目 https://[github.com/yangjianxin1/Firefly](https://github.com/yangjianxin1/Firefly)
运行：
`torchrun --nproc_per_node=2 train_qlora.py \
    --train_args_file train_args/qlora/config.json `
其中config.json配置：
`{
    "output_dir": "output",
    "model_name_or_path": "models/Baichuan-7B/",
    "train_file": "data/train.json",
    "num_train_epochs": 1,
    "per_device_train_batch_size": 1,
    "gradient_accumulation_steps": 16,
    "learning_rate": 1e-3,
    "max_seq_length": 5500,
    "logging_steps": 20,
    "save_steps": 50,
    "save_total_limit": 1,
    "lr_scheduler_type": "constant_with_warmup",
    "warmup_steps": 3000,
    "lora_rank": 64,
    "lora_alpha": 16,
    "lora_dropout": 0.05,
    "gradient_checkpointing": true,
    "disable_tqdm": false,
    "optim": "adamw_torch",
    "seed": 42,
    "fp16": true,
    "report_to": "wandb",
    "dataloader_num_workers": 0,
    "save_strategy": "steps",
    "weight_decay": 0,
    "max_grad_norm": 0.3,
    "remove_unused_columns": false
}`

**Comment by LZHgrla** [comment]
1. xtuner 的 config 在默认情况下，会将多条数据拼接为一条长度到`max_length`的数据（`pack_to_max_length=True`），这意味着每个训练样本都是您所设置的 `max_length=6000`，对于模型而言这一长度太长了，可能会导致训练不收敛。因此，建议您可以尝试将 `pack_to_max_length` 设为 `False` 再次尝试。
2. 可以去掉 `deepspeed` 训练，判断是否是由 `deepspeed`引入的不收敛。

**Comment by lionelmessigoat** [comment]
我代码搞错了

**Comment by lionelmessigoat** [comment]
> 1. xtuner 的 config 在默认情况下，会将多条数据拼接为一条长度到`max_length`的数据（`pack_to_max_length=True`），这意味着每个训练样本都是您所设置的 `max_length=6000`，对于模型而言这一长度太长了，可能会导致训练不收敛。因此，建议您可以尝试将 `pack_to_max_length` 设为 `False` 再次尝试。
> 2. 可以去掉 `deepspeed` 训练，判断是否是由 `deepspeed`引入的不收敛。

我代码搞错了
